docs/user-guide/database/pagination.rst

```diff
@@ -0,0 +1,229 @@
+
+If the model includes any timestamps, the model validation must be able to convert them from the time format stored in the database (see :doc:`datetime`) to an appropriate Python `~datetime.datetime`.
+The easiest way to do this is to declare those fields as having the `safir.pydantic.UtcDatetime` type.
+See :ref:`pydantic-datetime` for more information.
+
+Run the query
+-------------
+
+Finally, you can run the query.
+There are two ways to do this depending on how the query is structured.
+
+If the SQL query returns a single ORM model for each result row, use `~safir.database.PaginatedQueryRunner.query_object`:
+
+.. code-block:: python
+
+   results = await runner.query_object(
+       session, stmt, cursor=cursor, limit=limit
+   )
+
+If the SQL query returns a tuple of individually selected attributes that correspond to the fields of the result model (the first parameter to the `~safir.database.PaginatedQueryRunner` constructor), use `~safir.database.PaginatedQueryRunner.query_row`:
+
+.. code-block:: python
+
+   results = await runner.query_row(session, stmt, cursor=cursor, limit=limit)
+
+Either way, the results will be a `~safir.database.PaginatedList` wrapping a list of Pydantic models of the appropriate type.
+
+Returning paginated results
+===========================
+
+HTTP provides the ``Link`` header (:rfc:`8288`) to declare relationships between multiple web responses.
+Using a ``Link`` header with relation types ``first``, ``next``, and ``prev`` is a standard way of providing the client with pagination information.
+
+The Safir `~safir.database.PaginatedList` type provides a method, `~safir.database.PaginatedList.link_header`, which returns the contents of an HTTP ``Link`` header for a given paginated result.
+It takes as its argument the base URL for the query (usually the current URL of a route handler).
+This is the recommended way to return pagination information alongside a result.
+
+Here is a very simplified example of a route handler that sets this header:
+
+.. code-block:: python
+
+   @router.get("/query", response_class=Model)
+   async def query(
+       *,
+       cursor: Annotated[
+           ModelCursor | None,
+           Query(),
+           BeforeValidator(lambda c: ModelCursor.from_str(c) if c else None),
+       ] = None,
+       limit: Annotated[int | None, Query()] = None,
+       session: Annotated[
+           async_scoped_session, Depends(db_session_dependency)
+       ],
+       request: Request,
+       response: Response,
+   ) -> list[Model]:
+       runner = PydanticQueryRunner(Model, ModelCursor)
+       stmt = build_query(...)
+       results = await runner.query_object(
+           session, stmt, cursor=cursor, limit=limit
+       )
+       if cursor or limit:
+           response.headers["Link"] = results.link_header(request.url)
+           response.headers["X-Total-Count"] = str(results.count)
```

Comment:
I'm not sure what different API you are thinking of here. Something like an `add_headers` method that takes the request? I'm a little leery of this because of the separation of concerns between the generic code and the web API; it feels a bit cleaner for the handler to decide what header name to use for the count, since that's not standardized.

This did make me realize that I should add separate methods to return the URLs to the first block of data, the next, and the previous to better support web applications that want to return the navigation links in a wrapper data structure around the results rather than in HTTP headers.

Reply:
I was thinking that there could be a `results.headers` that you could merge into `response.headers`, but I don't feel strongly, and you make a good point that the count header is not standardized.